Add Hibernate Search management endpoint (#35065)
quarkusio/quarkus: docs/src/main/asciidoc/hibernate-search-orm-elasticsearch.adoc

See link:{hibernate-search-orm-elasticsearch-aws-guide}#aws-configuration-reference[the documentation for the Hibernate Search ORM + Elasticsearch AWS extension]
for more information.

[[management]]
== Management endpoint

:extension-status: preview

> include::{includes}/extension-status.adoc[]

Reviewer: I'm not sure it makes sense to do this at the section level; it's not the whole extension we're talking about, but just one feature.

Maybe do something similar to what we did for outbox-polling coordination instead? See https://quarkus.io/guides/hibernate-search-orm-elasticsearch#coordination

> Hibernate Search extension has a capability to reindex the data through the xref:./management-interface-reference.adoc[management interface].

Reviewer: ```suggestion
The Hibernate Search extension provides an HTTP endpoint to reindex your data through the xref:./management-interface-reference.adoc[management interface].
```

By default, this endpoint is not available. It can be enabled through configuration properties as shown below.

[source,properties]
----
quarkus.management.enabled=true <1>
quarkus.hibernate-search.management.enabled=true <2>
----
<1> Enable the xref:./management-interface-reference.adoc[management interface].
<2> Enable Hibernate Search specific management endpoints.

Once the management is enabled, data can be re-indexed via `/q/reindex`, where `/q` is the default management root path.

> Reindex path (`/reindex`) can be changed via configuration property as shown below.

Reviewer: ```suggestion
The reindexing path (`/reindex`) can be changed via configuration property as shown below.
```

[source,properties]
----

> quarkus.hibernate-search.management.root-path=custom-root-path <1>

Reviewer: Feels weird that `/reindex` is called the "root" path... it would mean that if we ever add another endpoint to the management interface, it can only be named `/reindex/something`?

Wouldn't it make sense for the default to be something like `/q/hibernate-search/reindex` and for the "root path" to only change "hibernate-search" (e.g. it'd become `/q/custom-root-path/reindex`)?

----
<1> Use a custom `custom-root-path` path for reindex management endpoint.
If the default management root path is used then the reindex path becomes `/q/custom-root-path`.

This endpoint accepts `POST` requests with `application/json` content type only.

> All indexed entities will be re-indexed if an empty request body is submitted.

Reviewer: Maybe add a simple example here? Just so that people don't think they need to put all the complicated configuration below.

If only a subset of entities must be re-indexed or
if there is a need to have a custom configuration of the underlying mass indexer
then this information can be passed through the request body as shown below.

[source,json]
----
{
  "filter": { <1>
    "types": ["EntityName1", "EntityName2", "EntityName3", ...], <2>
    "tenants": ["tenant1", "tenant2", ...] <3>
  },
  "massIndexer":{ <4>
    "typesToIndexInParallel": 1, <5>
    "threadsToLoadObjects": 6,  <6>
    "batchSizeToLoadObjects": 10, <7>
    "cacheMode": "IGNORE", <8>
    "mergeSegmentsOnFinish": false, <9>
    "mergeSegmentsAfterPurge": true, <10>
    "dropAndCreateSchemaOnStart": false, <11>
    "purgeAllOnStart": true, <12>
    "idFetchSize": 100, <13>
    "transactionTimeout": 100000, <14>
  }
}
----

> <1> Filter object that allows to limit the amount of indexed entities.

Reviewer: ```suggestion
<1> Filter object that allows to limit the scope of reindexing.
```

> <2> An array of entity names that should be re-indexed, if empty -- all entities will be re-indexed.

Reviewer: ```suggestion
<2> An array of entity names that should be re-indexed. If unspecified or empty, all entity types will be re-indexed.
```

> <3> An array of tenant ids. In case of multi tenancy, re-indexing can be limited to a subset of available tenants.
> If empty data for all tenants will be re-indexed.

Reviewer: ```suggestion
<3> An array of tenant ids, in case of multi-tenancy. If unspecified or empty, all tenants will be re-indexed.
```

<4> Mass indexer configuration object.
<5> Sets the number of entity types to be indexed in parallel.
<6> Sets the number of threads to be used to load the root entities.
<7> Sets the batch size used to load the root entities.
<8> Sets the cache interaction mode for the data loading tasks.

> <9> If each index is merged into a single segment after indexing.
> <10> If each index is merged into a single segment after the initial index purge, just before indexing.
> <11> If the indexes and their schema (if they exist) should be dropped and re-created before indexing.
> <12> If all entities are removed from the indexes before indexing.

Reviewer: ```suggestion
<9> Whether each index is merged into a single segment after indexing.
<10> Whether each index is merged into a single segment after the initial index purge, just before indexing.
<11> Whether the indexes and their schema (if they exist) should be dropped and re-created before indexing.
<12> Whether all entities are removed from the indexes before indexing.
```

<13> Specifies the fetch size to be used when loading primary keys if objects to be indexed.
<14> Specifies the timeout of transactions for loading ids and entities to be re-indexed.
+
Note all the properties in the json are optional, and only those that are needed should be used.

For more detailed information on mass indexer configuration see the
link:{hibernate-search-docs-url}#indexing-massindexer-parameters[corresponding section of the Hibernate Search reference documentation].

> Submitting the reindexing request will trigger the indexing in the background.

Reviewer: ```suggestion
Submitting the reindexing request will trigger indexing in the background. Mass indexing progress will appear in the application logs.
```

For testing purposes, it might be useful to know when the indexing finished. Adding `blocking=true` query parameter to the URL

> will make mass indexer run in the blocking mode and response will not be returned until mass indexer finishes the process.

Reviewer: ```suggestion
will make mass indexer run in the blocking mode and the response will only be returned when mass indexer finishes the process.
```

> When working with multiple persistence units the name of the persistence unit to reindex can be supplied through the

Reviewer: ```suggestion
When working with multiple persistence units, the name of the persistence unit to reindex can be supplied through the
```
